File: June-LeetCoding-Challenge/17-Surrounded-Regions/Surrounded-Regions.cpp
const int dx[4] = {-1, 0, 1, 0};
const int dy[4] = {0, -1, 0, 1};

class Solution {
public:
    int n, m;
    void bfs(int x0, int y0, vector<vector<char>>& board, vector<vector<int>>& vis){
        queue<pair<int, int>> q;
        vis[x0][y0] = 1;
        q.push(make_pair(x0, y0));
        while(!q.empty()){
            int x = q.front().first, y = q.front().second;
            q.pop();
            for (int i = 0; i < 4; i++){
                int xx = x + dx[i];
                int yy = y + dy[i];
                if (xx >= 0 and xx < n and yy >= 0 and yy < m and board[xx][yy] == 'O' and vis[xx][yy] == 0){
                    vis[xx][yy] = 1;
                    q.push(make_pair(xx,yy));
                }
            }
            
        }
    }
    
    void solve(vector<vector<char>>& board) {
        n = board.size();
        if (n == 0) return;
        m = board[0].size();
        vector<vector<int>> vis(n, vector<int>(m, 0));
        for (int i = 0; i < n; i++) {
            if (board[i][0] == 'O' and vis[i][0] == 0)
                bfs(i, 0, board, vis);
            if (board[i][m-1] == 'O' and vis[i][m-1] == 0)
                bfs(i, m-1, board, vis);
        }
        for (int i = 0; i < m; i++) {
            if (board[0][i] == 'O' and vis[0][i] == 0)
                bfs(0, i, board, vis);
            if (board[n-1][i] == 'O' and vis[n-1][i] == 0)
                bfs(n-1, i, board, vis);
        }
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                if (board[i][j] == 'O' and vis[i][j] == 0) board[i][j] = 'X';
               
    }
};

File: June-LeetCoding-Challenge/17-Surrounded-Regions/Surrounded-Regions.kt
class SurroundedRegionsKotlin130 {
    fun solve(board: Array<CharArray>) {
        if (board.isEmpty()) {
            return
        }
        val maxX = board.size
        val maxY = board[0].size
        val dp = Array(maxX) { IntArray(maxY) { 0 } }
        for (index in board[0].indices) {
            if (board[0][index] == 'O') {
                dfs(board, dp, 0, index, maxX, maxY)
            }
            if (board[maxX - 1][index] == 'O') {
                dfs(board, dp, maxX - 1, index, maxX, maxY)
            }
        }
        for (index in board.indices) {
            if (board[index][0] == 'O') {
                dfs(board, dp, index, 0, maxX, maxY)
            }
            if (board[index][maxY - 1] == 'O') {
                dfs(board, dp, index, maxY - 1, maxX, maxY)
            }
        }
        for (i in 1..board.size - 2) {
            for (j in 1..board[0].size - 2) {
                if (board[i][j] == 'O' && dp[i][j] == 0) {
                    board[i][j] = 'X'
                }
            }
        }
    }

    private val deltaX = intArrayOf(0, 0, 1, -1)
    private val deltaY = intArrayOf(1, -1, 0, 0)

    private fun dfs(
        board: Array<CharArray>,
        dp: Array<IntArray>,
        x: Int,
        y: Int,
        maxX: Int,
        maxY: Int
    ) {
        if (dp[x][y] == 0 && board[x][y] == 'O') {
            dp[x][y] = 1
            for (index in deltaX.indices) {
                val nextX = x + deltaX[index]
                val nextY = y + deltaY[index]
                if (inBoardMinus1(nextX, nextY, maxX, maxY)) {
                    dfs(board, dp, nextX, nextY, maxX, maxY)
                }
            }
        }
    }

    private fun inBoardMinus1(x: Int, y: Int, maxX: Int, maxY: Int) = x > 0 && y > 0 && x < maxX - 1 && y < maxY - 1
    /*
    fun solve(board: Array<CharArray>) {
        if (board.isEmpty()) {
            return
        }
        val maxX = board.size
        val maxY = board[0].size
        val dp = Array(maxX) { IntArray(maxY) { 0 } }
        val queue: Queue<Pair<Int, Int>> = LinkedList()
        for (index in board[0].indices) {
            if (board[0][index] == 'O') {
                queue.offer(Pair(0, index))
            }
            if (board[maxX - 1][index] == 'O') {
                queue.offer(Pair(maxX - 1, index))
            }
        }
        for (index in board.indices) {
            if (board[index][0] == 'O') {
                queue.offer(Pair(index, 0))
            }
            if (board[index][maxY - 1] == 'O') {
                queue.offer(Pair(index, maxY - 1))
            }
        }
        val deltaX = intArrayOf(0, 0, 1, -1)
        val deltaY = intArrayOf(1, -1, 0, 0)
        while (queue.isNotEmpty()) {
            val current = queue.poll()
            val x = current.first
            val y = current.second
            if (dp[x][y] == 0 && board[x][y] == 'O') {
                dp[x][y] = 1
                for (index in deltaX.indices) {
                    val nextX = x + deltaX[index]
                    val nextY = y + deltaY[index]
                    if (inBoardMinus1(nextX, nextY, maxX, maxY)) {
                        queue.offer(Pair(nextX, nextY))
                    }
                }
            }
        }
        for (i in 1..board.size - 2) {
            for (j in 1..board[0].size - 2) {
                if (board[i][j] == 'O' && dp[i][j] == 0) {
                    board[i][j] = 'X'
                }
            }
        }
    }

    private fun inBoardMinus1(x: Int, y: Int, maxX: Int, maxY: Int) = x > 0 && y > 0 && x < maxX - 1 && y < maxY - 1

     */
}

fun main() {
    val solution = SurroundedRegionsKotlin130()
    val testArray = arrayOf(
        charArrayOf('X', 'X', 'X', 'X'),
        charArrayOf('X', 'O', 'O', 'X'),
        charArrayOf('X', 'X', 'O', 'X'),
        charArrayOf('X', 'O', 'X', 'X')
    )
    solution.solve(testArray)
    println()
}